File: data/src/main/java/home/oleg/placesnearme/data/repositories/FavoriteVenueRepositoryImpl.java
package home.oleg.placesnearme.data.repositories;

import java.util.List;

import home.oleg.placenearme.models.DetailedVenue;
import home.oleg.placenearme.repositories.FavoriteVenuesRepository;
import home.oleg.placesnearme.data.dao.DetailedVenueWithPhotosDao;
import home.oleg.placesnearme.data.mapper.DetailedVenueMapper;
import home.oleg.placesnearme.data.model.DetailedVenueDbEntity;
import home.oleg.placesnearme.data.model.DetailedVenueWithPhotos;
import io.reactivex.Completable;
import io.reactivex.Flowable;

public class FavoriteVenueRepositoryImpl implements FavoriteVenuesRepository {

    private final DetailedVenueWithPhotosDao dao;

    public FavoriteVenueRepositoryImpl(DetailedVenueWithPhotosDao dao) {
        this.dao = dao;
    }

    @Override
    public Flowable<List<DetailedVenue>> observeFavorites() {
        return dao.getAllFavorites().map(DetailedVenueMapper::map);
    }

    @Override
    public Completable addToFavorite(DetailedVenue venue) {
        return Completable.fromAction(() -> {
            DetailedVenueWithPhotos detailedVenueWithPhotos = DetailedVenueMapper.map(venue);
            detailedVenueWithPhotos.getVenue().setFavorite(true);
            dao.insert(detailedVenueWithPhotos);
        });
    }

    @Override
    public Completable deleteFromFavorite(DetailedVenue venue) {
        return Completable.fromAction(() -> {
            DetailedVenueWithPhotos detailedVenueWithPhotos = DetailedVenueMapper.map(venue);
            DetailedVenueDbEntity detailedVenueDbEntity = detailedVenueWithPhotos.getVenue();
            detailedVenueDbEntity.setFavorite(false);
            dao.update(detailedVenueDbEntity);
        });
    }

}

File: data/src/main/java/home/oleg/placesnearme/data/repositories/FavoriteVenueRepositoryImpl.kt
package home.oleg.placesnearme.data.repositories

import home.oleg.placenearme.models.DetailedVenue
import home.oleg.placenearme.repositories.FavoriteVenuesRepository
import home.oleg.placesnearme.data.dao.DetailedVenueDao
import home.oleg.placesnearme.data.mapper.DetailedVenueMapper
import home.oleg.placesnearme.data.model.DetailedVenueDbEntity
import home.oleg.placesnearme.data.model.DetailedVenueWithPhotos
import io.reactivex.Completable
import io.reactivex.Flowable
import javax.inject.Inject

class FavoriteVenueRepositoryImpl @Inject constructor(private val dao: DetailedVenueDao) : FavoriteVenuesRepository {

    override fun observeFavorites(): Flowable<List<DetailedVenue>> {
        return dao.allFavorites.map { DetailedVenueMapper.map(it) }
    }

    override fun addToFavorite(venue: DetailedVenue): Completable {
        return Completable.fromAction {
            val detailedVenueWithPhotos = DetailedVenueMapper.map(venue)
            detailedVenueWithPhotos.venue = detailedVenueWithPhotos.venue.copy(isFavorite = true)

            val oldDetailedVenueDbEntity = dao.getDetailedVenueById(venue.id)
            if (oldDetailedVenueDbEntity != null) {
                dao.update(detailedVenueWithPhotos.venue, detailedVenueWithPhotos.photos)
            } else {
                dao.insert(detailedVenueWithPhotos)
            }
        }
    }

    override fun deleteFromFavorite(venue: DetailedVenue): Completable {
        return Completable.fromAction {
            val (detailedVenueDbEntity) = DetailedVenueMapper.map(venue)
            dao.update(detailedVenueDbEntity.copy(isFavorite = false))
        }
    }

}
